akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -426,104 +424,273 @@ object StreamLayout {
-              case VirtualProcessor.Completed  ⇒ tryOnComplete(s)
-              case VirtualProcessor.Failed(ex) ⇒ tryOnError(s, ex)
-              case VirtualProcessor.Allowed    ⇒ // all good
-            }
-          } catch {
-            case NonFatal(ex) ⇒ sub.cancel()
-          }
+    @tailrec def rec(sub: Subscriber[Any]): Unit =
+      get() match {
+        case null => if (!compareAndSet(null, s)) rec(sub)
+        case subscription: Subscription =>
+          if (compareAndSet(subscription, Both(sub))) establishSubscription(sub, subscription)
+          else rec(sub)
+        case pub: Publisher[_] =>
+          if (compareAndSet(pub, Inert)) pub.subscribe(sub)
+          else rec(sub)
+        case _ =>
+          rejectAdditionalSubscriber(sub, "VirtualProcessor")
       }
+
+    if (s == null) {
+      val ex = subscriberMustNotBeNullException
+      try rec(Inert.subscriber)
+      finally throw ex // must throw NPE, rule 2:13
+    } else rec(s.asInstanceOf[Subscriber[Any]])
   }
 
-  override def onSubscribe(s: Subscription): Unit = {
-    requireNonNullSubscription(s)
-    val wrapped = new Sub(s)
-    if (subscriptionStatus.compareAndSet(null, wrapped)) () // wait for Subscriber
-    else
-      subscriptionStatus.get match {
-        case sub: Subscriber[_] ⇒
-          terminationStatus.get match {
-            case Allowed ⇒
-              /*
-               * There is a race condition here: if this thread reads the subscriptionStatus after
-               * set set() in subscribe() but then sees the terminationStatus before the getAndSet()
-               * is published then we will rely upon the downstream Subscriber for cancelling this
-               * Subscription. I only mention this because the TCK requires that we handle this here
-               * (since the manualSubscriber used there does not expose this behavior).
-               */
-              s.cancel()
-            case _ ⇒
-              tryOnSubscribe(sub, wrapped)
-              wrapped.closeLatch() // allow onNext only now
-              terminationStatus.set(Allowed)
+  override final def onSubscribe(s: Subscription): Unit = {
+    @tailrec def rec(obj: AnyRef): Unit =
+      get() match {
+        case null => if (!compareAndSet(null, obj)) rec(obj)
+        case subscriber: Subscriber[_] =>
+          obj match {
+            case subscription: Subscription =>
+              if (compareAndSet(subscriber, Both.create(subscriber))) establishSubscription(subscriber, subscription)
+              else rec(obj)
+            case pub: Publisher[_] =>
+              getAndSet(Inert) match {
+                case Inert => // nothing to be done
+                case _     => pub.subscribe(subscriber.asInstanceOf[Subscriber[Any]])
+              }
           }
-        case sub: Subscription ⇒
-          s.cancel() // reject further Subscriptions
+        case _ =>
+          // spec violation
+          tryCancel(s)
```

Comment:
hm, but what if we got the onSubscribe in state `Both`, shouldn't we fail our downstream (who is innocent in this case)?